cpp/include/types.h

```diff
@@ -19,13 +19,16 @@ void gdf_col_delete(gdf_column* col);
 
 void gdf_col_release(gdf_column* col);
 
+typedef enum gdf_prop_type{GDF_PROP_UNDEF, GDF_PROP_FALSE, GDF_PROP_TRUE} GDFPropType;
+
 struct gdf_graph_properties {
   bool directed;
   bool weighted;
   bool multigraph;
   bool bipartite;
   bool tree;
-  gdf_graph_properties() : directed(false), weighted(false), multigraph(false), bipartite(false), tree(false){}
+  GDFPropType has_negative_edges;
```

Comment:
Does has_negative_edges need to be gdf_graph_properties's member variable?

Currently, this is used only for SSSP, but we are running has_negative_val() for every gdf_add_adj_list() call. This incurs additional performance overhead to every graph algorithm (and currently, cuGraph C++ API has two mechanisms to initialize graphs (add_edge_list_view and add_adj_list_view) and three mechanisms to convert graph formats (add_edge_list, add_adj_list, and add_transposed_adj_list... I know the names are misnomers and can be very confusing... We're planning to fix the gdf_graph structure...). If we are going to update has_negative_edges for every graph object, we need to consider all these possibilities, but I think this will add too much performance overhead for other graph analytics if they do not use has_negative_edges.

An alternative is too call has_negative_val() inside the SSSP routine; this adds unnecessary overhead if we call SSSP or other algorithms using has_negative_edges on the same graph multiple times but I think this is less overhead than incurring the overhead of has_negative_val() for every graph analytics.

Reply:
I had the exact same discussion with @afender yesterday. I originally had it inside the SSSP code behind a flag, and my plan was to use some other persistent storage to set the flag on first run but not on subsequent runs. That makes it somewhat kludgy though.

Reply:
OK, may be one way to resolve this is (and you already defined GDF_PROP_UNDEF) to maintain this as a some optional graph state variable which may be set to UNDEFINED in most cases, but at the beginning of any algorithm requiring this property, you may check whether it is UNDEFINED, then run has_negative_val)() and set this value. If set, then, just reuse.

If we do this, we don't need to run has_negative_val() everytime we initialize graphs.

We will consider doing this in a more elegant way in the planning gdf_graph redesign, but as a temporary solution, you may update this in the SSSP call instead of add_adj_list().

Reply:
Yes, fair point. I'll move the negative edges check inside sssp.

Reply:
So, in the long run, we want a safe mode where the graph input is validated and a performance mode where we skip these checks. This should be controlled by the cugraph handle. Having these checks in one place at loading time allows preserving the analytics call performance even when running in safe mode.

I agree that most cugraph algorithms don't care about negative weights but for the majority that's because they just don't use weights at all. In this case, this check does not happen and performance is not impacted.

Reply:
And don't forget to reset has_negative_edge to UNDEFINED when the graph is reinitialized (gdf_edge_list_view & gdf_adj_list_view).

Reply:
Yeah... I think Alex's comment makes sense. Then, we should make sure has_negative_edge is properly updated in every possible path of graph initialization (not just add_adj_list which coverts an existing graph to the CSR format).

Reply:
For completeness, spectral clustering and Louvain are other algorithms that would benefit from that since they support positive weights only in theory. I think other algo don't use weights at all.

Reply:
We are using CSR for (AFAIK) every graph algorithm (minus PageRank? which uses CSC instead), so doing this in add_adj_list is safe for most cases, but I don't think it's future proof.

Reply:
Agreed. I thought it was in gdf_edge_list_view actually. It is probably better to add it for the edge list since this is supposed to be the place where we first encounter the input (in data science workloads).

Reply:
I have moved the checks to adj and edge list view functions now. So they should be called on all normal paths.